cpp/src/plasma/io/basic_connection.h

```diff
@@ -0,0 +1,136 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PLASMA_BASIC_CONNECTION_H
+#define PLASMA_BASIC_CONNECTION_H
+
+#ifndef ASIO_STANDALONE
+#define ASIO_STANDALONE
+#endif
+
+#include <deque>
+#include <memory>
+#include <string>
+#include <system_error>
+#include <utility>
+#include <vector>
+
+#include "asio.hpp"  // NOLINT
+
+namespace plasma {
+namespace io {
+
+using AsyncWriteCallback = std::function<void(const std::error_code&)>;
+// TODO(suquark): Change it according to the platform.
+using PLASMA_STREAM_TYPE = asio::basic_stream_socket<asio::local::stream_protocol>;
+using PLASMA_ACCEPTOR_TYPE = asio::local::stream_protocol::acceptor;
+
+/// Create a local acceptor depends on the platform.
+PLASMA_ACCEPTOR_TYPE CreateLocalAcceptor(asio::io_context& io_context,
+                                         const std::string& name);
+
+/// Create a local stream depends on the platform.
+PLASMA_STREAM_TYPE CreateLocalStream(asio::io_context& io_context,
+                                     const std::string& name);
+
+/// A message that is queued for writing asynchronously.
+struct AsyncWriteBuffer {
+  virtual void ToBuffers(std::vector<asio::const_buffer>& message_buffers) {}
+  AsyncWriteCallback handler;
+  virtual ~AsyncWriteBuffer() {}
+};
+
+template <typename T>
+class BasicConnection : public std::enable_shared_from_this<BasicConnection<T>> {
```

Comment:
I think there is too many connections here right now. Let's just have `Connection` (which is `BasicConnection`) and `ServerConnection` (which is both PlasmaServerConnection and the current ServerConnection), as well as `ClientConnection` (which is both PlasmaClientConnection and the current ClientConnection). This will also make the code elsewhere more readable, since then we can drop the `Plasma` prefix from all the connection, which is understood since we are in the `plasma` namespace.